File: georocket-server/src/main/java/io/georocket/api/index/IndexerFactory.java
package io.georocket.api.index;

import java.util.Map;

import io.georocket.api.query.QueryCompiler;

/**
 * Factory for {@link Indexer} objects
 * @author Michel Kraemer
 */
public interface IndexerFactory extends QueryCompiler {
  /**
   * @return the Elasticsearch mapping for the results returned by the indexer
   */
  Map<String, Object> getMapping();
  
  /**
   * @return a new instance of {@link Indexer}
   */
  Indexer createIndexer();
}

File: kotlin_impl/src/main/kotlin/io/georocket/index/IndexerFactory.kt
package io.georocket.index

import io.georocket.index.generic.BoundingBoxIndexerFactory
import io.georocket.index.generic.GenericAttributeIndexerFactory
import io.georocket.index.geojson.GeoJsonIdIndexerFactory
import io.georocket.index.xml.GmlIdIndexerFactory
import io.georocket.index.xml.XalAddressIndexerFactory
import io.georocket.query.QueryCompiler
import io.georocket.util.StreamEvent

/**
 * Factory for [Indexer] objects
 * @author Michel Kraemer
 */
interface IndexerFactory : QueryCompiler {
  companion object {
    val ALL = listOf(
      BoundingBoxIndexerFactory(),
      GenericAttributeIndexerFactory(),
      GmlIdIndexerFactory(),
      GeoJsonIdIndexerFactory(),
      XalAddressIndexerFactory()
    )
  }

  /**
   * Returns a new instance of [Indexer] for the given [eventType]. Returns
   * `null` if the type is not supported.
   */
  fun <T : StreamEvent> createIndexer(eventType: Class<T>): Indexer<T>?
}
